File: backend/app/modules/auth/permissions.py
"""RBAC 权限常量与预置角色定义。"""


class Permissions:
    # 项目
    PROJECT_CREATE = "project:create"
    PROJECT_EDIT = "project:edit"
    PROJECT_DELETE = "project:delete"
    PROJECT_VIEW = "project:view"
    # 需求
    REQUIREMENT_UPLOAD = "requirement:upload"
    REQUIREMENT_DELETE = "requirement:delete"
    REQUIREMENT_REVIEW = "requirement:review"
    REQUIREMENT_VIEW = "requirement:view"
    # 用例
    TESTCASE_CREATE = "testcase:create"
    TESTCASE_EDIT = "testcase:edit"
    TESTCASE_DELETE = "testcase:delete"
    TESTCASE_VIEW = "testcase:view"
    TESTCASE_GENERATE = "testcase:generate"
    TESTCASE_APPROVE = "testcase:approve"
    # LLM
    LLM_CONFIG = "llm:config"
    LLM_CHAT = "llm:chat"
    # UI 自动化（二期）
    UI_ENV_VIEW = "ui_env:view"
    UI_ENV_CREATE = "ui_env:create"
    UI_ENV_EDIT = "ui_env:edit"
    UI_ENV_DELETE = "ui_env:delete"
    # UI 自动化执行（二期 Task 9.6）
    # 拆三个权限是为了让"看历史"和"按下立即执行"是不同的责任：
    # tester 默认能 RUN，viewer 只能 VIEW；STOP 单独拎出来留作未来"只让发起人能停"的策略升级位
    UI_EXEC_VIEW = "ui_exec:view"
    UI_EXEC_RUN = "ui_exec:run"
    UI_EXEC_STOP = "ui_exec:stop"
    # Task 9.7：调试模式（POST /continue 推进单步）。
    # 与 RUN 拆开是为了"灰度放给资深 QA"——调试会卡住浏览器实例 30 分钟，
    # 不希望随便谁都能开
    UI_EXEC_DEBUG = "ui_exec:debug"
    # 接口测试（放在 UI 自动化菜单下，独立于浏览器执行权限）
    API_TEST_VIEW = "api_test:view"
    API_TEST_EDIT = "api_test:edit"
    API_TEST_RUN = "api_test:run"
    # 测试物料（二期 Task 8.5）
    TEST_DATA_VIEW = "test_data:view"
    TEST_DATA_EDIT = "test_data:edit"
    TEST_DATA_REVEAL = "test_data:reveal"
    TEST_DATA_IMPORT = "test_data:import"
    # 提示词管理（2026-05 拆分）
    # 历史上 ``prompts/router.py`` 复用 ``REQUIREMENT_*`` 权限——语义混乱：
    # 改一段 LLM prompt 模板和上传需求文档完全是两个责任。这里独立出来，
    # 同时让前端 ``RoleEditDialog`` 的菜单树能展示"提示词管理"这一项
    # （之前用户编辑角色时找不到这个菜单的开关）。``init_data._seed_roles``
    # 每次启动自动同步系统角色权限，所以已部署的 admin / project_manager
    # 重启后会自动补上新权限，**无需手写 alembic 迁移**。
    PROMPT_VIEW = "prompt:view"
    PROMPT_EDIT = "prompt:edit"
    PROMPT_DELETE = "prompt:delete"
    # Skill 技能包（Phase 12 Task 12.4）
    SKILL_VIEW = "skill:view"
    SKILL_EDIT = "skill:edit"
    SKILL_DELETE = "skill:delete"
    SKILL_IMPORT = "skill:import"
    SKILL_EXPORT = "skill:export"
    SKILL_SCAN = "skill:scan"
    SKILL_CHAT_ACTIVATE = "skill:chat_activate"
    # 管理
    USER_MANAGE = "user:manage"
    ROLE_MANAGE = "role:manage"


ALL_PERMISSIONS = sorted([
    v for k, v in vars(Permissions).items() if not k.startswith("_")
])

SYSTEM_ROLES: dict[str, dict] = {
    "admin": {
        "display_name": "管理员",
        "description": "拥有所有权限",
        "permissions": ALL_PERMISSIONS,
    },
    "project_manager": {
        "display_name": "项目经理",
        "description": "管理项目、需求和用例，使用 AI 对话",
        "permissions": [
            Permissions.PROJECT_CREATE,
            Permissions.PROJECT_EDIT,
            Permissions.PROJECT_DELETE,
            Permissions.PROJECT_VIEW,
            Permissions.REQUIREMENT_UPLOAD,
            Permissions.REQUIREMENT_DELETE,
            Permissions.REQUIREMENT_REVIEW,
            Permissions.REQUIREMENT_VIEW,
            Permissions.TESTCASE_CREATE,
            Permissions.TESTCASE_EDIT,
            Permissions.TESTCASE_DELETE,
            Permissions.TESTCASE_VIEW,
            Permissions.TESTCASE_GENERATE,
            Permissions.TESTCASE_APPROVE,
            Permissions.LLM_CHAT,
            Permissions.UI_ENV_VIEW,
            Permissions.UI_ENV_CREATE,
            Permissions.UI_ENV_EDIT,
            Permissions.UI_ENV_DELETE,
            Permissions.UI_EXEC_VIEW,
            Permissions.UI_EXEC_RUN,
            Permissions.UI_EXEC_STOP,
            Permissions.UI_EXEC_DEBUG,
            Permissions.API_TEST_VIEW,
            Permissions.API_TEST_EDIT,
            Permissions.API_TEST_RUN,
            Permissions.TEST_DATA_VIEW,
            Permissions.TEST_DATA_EDIT,
            Permissions.TEST_DATA_REVEAL,
            Permissions.TEST_DATA_IMPORT,
            Permissions.SKILL_VIEW,
            Permissions.SKILL_EDIT,
            Permissions.SKILL_DELETE,
            Permissions.SKILL_IMPORT,
            Permissions.SKILL_EXPORT,
            Permissions.SKILL_SCAN,
            Permissions.SKILL_CHAT_ACTIVATE,
            # 项目经理需要为团队定制提示词模板（如评审风格、用例生成口吻）
            Permissions.PROMPT_VIEW,
            Permissions.PROMPT_EDIT,
            Permissions.PROMPT_DELETE,
        ],
    },
    "tester": {
        "display_name": "测试人员",
        "description": "查看项目和需求，管理用例，使用 AI 对话",
        "permissions": [
            Permissions.PROJECT_VIEW,
            Permissions.REQUIREMENT_VIEW,
            Permissions.TESTCASE_CREATE,
            Permissions.TESTCASE_EDIT,
            Permissions.TESTCASE_DELETE,
            Permissions.TESTCASE_VIEW,
            Permissions.TESTCASE_GENERATE,
            Permissions.TESTCASE_APPROVE,
            Permissions.LLM_CHAT,
            Permissions.UI_ENV_VIEW,
            Permissions.UI_EXEC_VIEW,
            Permissions.UI_EXEC_RUN,
            Permissions.UI_EXEC_STOP,
            Permissions.UI_EXEC_DEBUG,
            Permissions.API_TEST_VIEW,
            Permissions.API_TEST_EDIT,
            Permissions.API_TEST_RUN,
            Permissions.TEST_DATA_VIEW,
            Permissions.TEST_DATA_EDIT,
            Permissions.TEST_DATA_IMPORT,
            Permissions.SKILL_VIEW,
            Permissions.SKILL_EDIT,
            Permissions.SKILL_IMPORT,
            Permissions.SKILL_EXPORT,
            Permissions.SKILL_SCAN,
            Permissions.SKILL_CHAT_ACTIVATE,
            # 测试人员只读提示词（生成用例 / 评审走的是项目内已有模板，不允许
            # 改模板免得影响他人）
            Permissions.PROMPT_VIEW,
        ],
    },
    "viewer": {
        "display_name": "只读用户",
        "description": "仅可查看项目、需求和用例",
        "permissions": [
            Permissions.PROJECT_VIEW,
            Permissions.REQUIREMENT_VIEW,
            Permissions.TESTCASE_VIEW,
            Permissions.UI_ENV_VIEW,
            Permissions.UI_EXEC_VIEW,
            Permissions.API_TEST_VIEW,
            Permissions.TEST_DATA_VIEW,
            Permissions.PROMPT_VIEW,
            Permissions.SKILL_VIEW,
        ],
    },
}

File: frontend/src/constants/permissions.ts
/**
 * 权限元数据：把后端权限 key 按"菜单"维度组织成中文树形结构。
 * - 菜单节点本身（key = 形如 "menu:<scope>"）作为「赋予该菜单全部权限」的快捷开关；
 *   勾选菜单 = 自动选中菜单下所有子权限；不勾选 = 该角色看不到此菜单。
 * - 子节点直接对应后端 Permissions.* 字符串。
 */

export interface PermissionLeaf {
  key: string;
  label: string;
  description?: string;
}

export interface PermissionGroup {
  /** 内部分组 key（用于"勾选菜单 = 全选子项"的本地标记，不会下发到后端） */
  key: string;
  /** 菜单中文名 */
  label: string;
  /** 关联前端导航菜单（路由 name），勾选则该角色可见此菜单 */
  routeNames?: string[];
  /** 描述 */
  description?: string;
  /** 子权限列表（后端 Permissions 枚举值） */
  permissions: PermissionLeaf[];
}

export const PERMISSION_GROUPS: PermissionGroup[] = [
  {
    key: "menu:project",
    label: "项目管理",
    routeNames: ["ProjectList", "ProjectSettings"],
    description: "管理产品 / 业务项目空间及成员",
    permissions: [
      { key: "project:view", label: "查看项目", description: "浏览项目列表与详情" },
      { key: "project:create", label: "创建项目" },
      { key: "project:edit", label: "编辑项目", description: "修改项目信息、成员等" },
      { key: "project:delete", label: "删除项目" },
    ],
  },
  {
    key: "menu:requirement",
    label: "需求管理",
    routeNames: ["RequirementList", "RequirementDetail"],
    description: "上传需求文档并使用 AI 评审",
    permissions: [
      { key: "requirement:view", label: "查看需求文档" },
      { key: "requirement:upload", label: "上传文档" },
      { key: "requirement:delete", label: "删除文档" },
      { key: "requirement:review", label: "发起 AI 评审" },
    ],
  },
  {
    key: "menu:testcase",
    label: "测试用例",
    routeNames: ["TestcaseList"],
    description: "组织、编辑及生成测试用例",
    permissions: [
      { key: "testcase:view", label: "查看用例" },
      { key: "testcase:create", label: "新建用例" },
      { key: "testcase:edit", label: "编辑用例" },
      { key: "testcase:delete", label: "删除用例" },
      { key: "testcase:generate", label: "AI 生成用例" },
      { key: "testcase:approve", label: "审核 / 接受用例" },
    ],
  },
  {
    key: "menu:chat",
    label: "AI 对话",
    routeNames: ["AIChat"],
    description: "使用平台内置的 AI 对话能力",
    permissions: [{ key: "llm:chat", label: "使用 AI 对话" }],
  },
  {
    key: "menu:ui-automation",
    label: "UI 自动化",
    routeNames: [
      "UIEnvironmentList",
      "UIEnvironmentListForProject",
      "UIExecutionHistory",
      "UIExecutionHistoryGlobal",
      "UIExecutionDetail",
      "UIExecutionMonitor",
    ],
    description: "维护 UI 执行环境、登录态、前置步骤与自动化执行记录",
    permissions: [
      { key: "ui_env:view", label: "查看 UI 环境" },
      { key: "ui_env:create", label: "新建 UI 环境" },
      {
        key: "ui_env:edit",
        label: "编辑 UI 环境",
        description: "包括前置步骤、凭据、登录态清除和前置步骤试跑",
      },
      { key: "ui_env:delete", label: "删除 UI 环境" },
      {
        key: "ui_exec:view",
        label: "查看 UI 执行记录",
        description: "查看执行历史、执行详情、实时监控、回放、视频、Trace 和截图",
      },
      {
        key: "ui_exec:run",
        label: "执行 UI 测试",
        description: "从用例管理发起执行、预检、重跑失败用例和按历史配置重跑",
      },
      {
        key: "ui_exec:stop",
        label: "停止 / 删除 UI 执行",
        description: "停止执行中任务，以及物理删除执行记录和关联截图、视频、Trace",
      },
      {
        key: "ui_exec:debug",
        label: "调试 UI 执行",
        description: "使用调试模式并在暂停时继续下一步",
      },
    ],
  },
  {
    key: "menu:api-test",
    label: "API 管理",
    routeNames: [
      "ApiEnvironmentList",
      "ApiEnvironmentListForProject",
      "ApiTestCases",
      "ApiTestCasesGlobal",
      "ApiAutomationTasks",
      "ApiAutomationTasksGlobal",
    ],
    description: "维护 API 环境、API 列表、批量执行报告和 API 自动化任务",
    permissions: [
      {
        key: "api_test:view",
        label: "查看 API 管理",
        description: "查看环境配置、环境变量、API 列表、批量报告、API 自动化任务和执行历史",
      },
      {
        key: "api_test:edit",
        label: "编辑 API 管理",
        description: "创建 / 修改 / 删除 API 模块、环境、变量、接口和自动化任务",
      },
      {
        key: "api_test:run",
        label: "执行 API 测试",
        description: "单接口调试、批量执行、执行 API 自动化任务",
      },
    ],
  },
  {
    key: "menu:test-data",
    label: "测试物料",
    routeNames: ["TestDataView", "TestDataViewGlobal", "TestDataSetEditor"],
    description: "管理可复用的测试数据：账号、文件、参数化数据集",
    permissions: [
      { key: "test_data:view", label: "查看物料集" },
      { key: "test_data:edit", label: "编辑物料集", description: "创建 / 更新 / 删除集合与条目（不含文件上传）" },
      { key: "test_data:import", label: "批量导入与文件上传", description: "CSV / JSON 导入、物料文件上传、克隆物料集" },
      { key: "test_data:reveal", label: "查看密文明文", description: "读取 secret 类型物料的明文（每次调用都记录审计日志）" },
    ],
  },
  {
    key: "menu:llm-config",
    label: "LLM 配置",
    routeNames: ["LLMConfig"],
    description: "管理大模型供应商与默认参数",
    permissions: [
      { key: "llm:config", label: "管理 LLM 配置", description: "增删改与测试连接" },
    ],
  },
  {
    // 历史上提示词路由复用 REQUIREMENT_* 权限，导致这里没有独立分组——
    // 用户编辑角色时找不到"提示词管理"开关（2026-05 验收反馈）。后端已经
    // 拆出 ``PROMPT_*`` 三个权限并通过 ``init_data._seed_roles`` 自动下发
    // 到系统角色；前端这里同步加分组，让角色权限树能展示。
    key: "menu:prompt",
    label: "提示词管理",
    routeNames: ["PromptManagement"],
    description: "维护项目级 LLM 提示词模板与版本历史",
    permissions: [
      { key: "prompt:view", label: "查看提示词", description: "浏览模板列表 / 详情 / 版本历史" },
      {
        key: "prompt:edit",
        label: "编辑提示词",
        description: "新建 / 修改 / 设为默认 / 初始化内置模板",
      },
      { key: "prompt:delete", label: "删除提示词", description: "系统内置模板不可删除" },
    ],
  },
  {
    // Phase 12 Task 12.5：技能包独立菜单。后端 7 个 skill:* 权限通过
    // ``init_data._seed_roles`` 自动同步到系统角色，前端这里同步暴露，
    // 让 RoleEditDialog 能勾选"技能包管理"开关。
    key: "menu:skill",
    label: "技能包管理",
    routeNames: ["SkillManagement", "SkillEditor", "SkillUsageStats"],
    description: "管理项目技能包：导入 / 导出 / 编辑 / 安全扫描 / Chat 激活 / 使用统计",
    permissions: [
      { key: "skill:view", label: "查看技能", description: "浏览列表 / 详情 / 版本 / 使用统计" },
      { key: "skill:edit", label: "编辑技能", description: "新建 / 修改 / 启停（自动重新扫描）" },
      { key: "skill:delete", label: "删除技能", description: "内置技能不可删除" },
      { key: "skill:import", label: "导入技能", description: "ZIP 上传 / URL 拉取" },
      { key: "skill:export", label: "导出技能", description: "下载 OpenClaw 兼容 ZIP" },
      { key: "skill:scan", label: "重新安全扫描" },
      { key: "skill:chat_activate", label: "对话内手动激活", description: "在 chat 中选中 manual 技能" },
    ],
  },
  {
    key: "menu:user",
    label: "用户管理",
    routeNames: ["UserManagement"],
    description: "管理平台账户、状态与角色绑定",
    permissions: [{ key: "user:manage", label: "管理用户", description: "查看 / 新建 / 编辑 / 删除 / 改密" }],
  },
  {
    key: "menu:role",
    label: "角色管理",
    routeNames: ["RoleManagement"],
    description: "维护角色及其权限",
    permissions: [{ key: "role:manage", label: "管理角色", description: "新建 / 编辑 / 删除角色与权限分配" }],
  },
];

/**
 * 全部权限 key 的中文映射，用于角色列表里把英文权限渲染成中文短标签。
 */
export const PERMISSION_LABEL_MAP: Record<string, string> = (() => {
  const map: Record<string, string> = {};
  for (const group of PERMISSION_GROUPS) {
    for (const p of group.permissions) {
      map[p.key] = `${group.label} / ${p.label}`;
    }
  }
  return map;
})();

/**
 * 将一组权限 key 转换为"菜单维度"的中文标签数组（去重，按菜单分组聚合）。
 * 用于角色列表 / 详情的权限简洁展示。
 */
export function summarizePermissions(permissionKeys: string[]): { menu: string; full: boolean; count: number; total: number }[] {
  const set = new Set(permissionKeys);
  return PERMISSION_GROUPS.map((g) => {
    const total = g.permissions.length;
    const count = g.permissions.filter((p) => set.has(p.key)).length;
    return { menu: g.label, full: count === total && total > 0, count, total };
  }).filter((s) => s.count > 0);
}
